gorm/transaction.go

```diff
@@ -139,25 +133,42 @@ func (t *Transaction) Rollback() error {
 	if t.current == nil {
 		return nil
 	}
-	if reflect.ValueOf(t.current.CommonDB()).IsNil() {
-		return status.Error(codes.Unavailable, "Database connection not available")
+	sqlDB, err := t.current.DB()
+	if err == nil && reflect.ValueOf(sqlDB).IsNil() {
+		err = errors.New("no underlying connection")
 	}
-	t.current.Rollback()
-	err := t.current.Error
+	if err != nil {
+		return databaseConnectionError(err)
+	}
+
+	t.current = t.current.Rollback()
+	err = t.current.Error
 	t.current = nil
+
+	if err != nil && strings.Contains(err.Error(), "database is closed") {
+		return databaseConnectionError(err)
+	}
+
 	return err
 }
 
+func databaseConnectionError(err error) error {
+	return status.Errorf(codes.Unavailable, "Database connection not available: %v", err)
+}
+
 // Commit finishes transaction by calling `*gorm.DB.Commit()`
 // Reset current transaction and returns an error if any.
 func (t *Transaction) Commit(ctx context.Context) error {
 	t.mu.Lock()
 	defer t.mu.Unlock()
 
-	if t.current == nil || reflect.ValueOf(t.current.CommonDB()).IsNil() {
+	if t.current == nil {
+		return nil
+	}
+	if sqlDB, err := t.current.DB(); err == nil && reflect.ValueOf(sqlDB).IsNil() {
```

Comment:
This seems to have a new potential `err` case that is just dropped. Probably not a problem, but it doesn't look very clean

Reply:
Fair enough, changed